New config simple exp function

shangweilin opened the pull request:
Please help to review this branch.
The current version include the following semantics based on the new configuration:
1. Basic types (boolean and integers)
2. Basic expressions and statements (not including loops)
3. External and internal function calls
4. Basic testing engine

dwightguth commented on configuration.k:
```diff
@@ -0,0 +1,220 @@
+            // (see the <account> cell below for more notes about this)
+
+            <contracts>
+                <Contract multiplicity = "*">
```
this should be `type="Map"`

shangweilin replied:
OK, I will revise it accordingly.

shangweilin replied:
I just found that we define `<Contract>` as a "bag" instead of a "map". If I add `type="Mao"`, there will be errors as follows:
```[Error] Critical: failed to concatenate maps with common keys:```

dwightguth commented on configuration.k:
```diff
@@ -0,0 +1,220 @@
+                    <cname> .K </cname>
+                    
+                    <functions>
+                        <Function multiplicity="*">
```
as should this.

shangweilin replied:
OK, I will revise it accordingly.

dwightguth commented on configuration.k:
```diff
@@ -0,0 +1,220 @@
+            <activeAccounts> .Set </activeAccounts>
+	        <cntAccount> 0:Int </cntAccount>
+            <accounts>
+                <account multiplicity="*">
```
and this

shangweilin replied:
OK, I will revise it accordingly.

shangweilin replied:
I just found that we define `<account>` as a "bag" instead of a "map". If I add `type="Map"`, there will be errors as follows:
```[Error] Critical: failed to concatenate maps with common keys:```

shangweilin replied:
If I change add `type="Map"` to the `<account>`, even the following example has the same error:
`"code" : #solidity( new C(); )`

And, I found that the problem comes from the following rule:
```
rule [Allocate-State-Variables]:
    <k> #allocate(N, #varInfo(X:Id,E:Value,T:TypeName, #storage)) => . ...</k>
    <account>
        <acctID> #account(N,_:Id) </acctID>
        <context> CONTEXT:Map => CONTEXT[X <- #storedVar(!N:Int, T, #storage)] </context>
        <Storage> STORAGE:Map => STORAGE[!N <- E] </Storage>
        ...
    </account>
```

After applying the above rule, the same error happens:
[Error] Critical: failed to concatenate maps with common keys:

```
[(
    <acctID>(#account(Int(#"0"),, Id(#"C"))),
    <account>(
        <acctID>(#account(Int(#"0"),, Id(#"C"))),,
        <balance>(Int(#"0")),, 
        <context>(Id(#"a") |-> #storedVar(Int(#"1"),,int_SOLIDITY-SYNTAX(.KList),, #storage_SOLIDITY-SYNTAX(.KList))   ),,
        <code>(
            <associatedContract>(#EmptyK()),, 
            <availableContracts>(#EmptyK())),,
        <Storage>(Int(#"1") |-> #undef_Value_SOLIDITY-SYNTAX(.KList)   ),,
        <nonce>(Int(#"0"))),
    
    <account>(
        <acctID>(#account(Int(#"0"),, Id(#"C"))),,
        <balance>(Int(#"0")),, 
        <context>(.Map),,
        <code>(
            <associatedContract>(#EmptyK()),, 
            <availableContracts>(#EmptyK())),,
        <Storage>(.Map),, 
        <nonce>(Int(#"0")))
)]

  while evaluating function _AccountCellMap_
```

tomtomjhj replied:
@shangweilin @dfilaretti  I can't reproduce this. How did you get those results? I'm using the most recent version of K.  `krun tests/functionCall.sol` ends with `<mode> "SUCCESS" </mode>` and the other cells also look OK. BTW how did you get the step number?

shangweilin replied:
@tomtomjhj, to reproduce the error, you need to do the following changing of code:
1. Change Line 149 of `configuration.k` as `<account multiplicity="*" type="Map">`
2. Change Line 49 of `functionCall.sol` as
```
"code" : #solidity( new C();)
```

I got the step number by running the following command, where `i` from 1 to 37
`krun test/functionCall.sol --depth i`
The error happens from steps 37 to 38.

tomtomjhj replied:
I looked into the implementation of K builtin Map and kompiled rules. This is how a Map element rewriting is done:
1. look up the key, 
2. remove this key-value pair from the Map
3. update the value of this entry
4. concatenate this updated entry with the Map (without this key) from 2.

The key of the Map is `<acctID>` cell and its contents. In your rule [Allocate-State-Variables], the key is `<acctID> #account(N,_:Id) </acctID>` containing a K variable `_:Id`, which is an entity that has no relationship with any concrete terms before it is unified to something else.
The problem here is that the actual key in the map `<acctID> #account( 0, C) </acctID>` is different from the key your rule looks up for because `_:Id` is not `C`. Therefore, the lookup of the first step fails, no key-value pair removed, then finally, the concatenation of Maps with common key.

To resolve this, I would only put the account Id on the `<acctID>` cell and devise a way to get the contract name when needed instead of using `#account(..)`.
e.g. new `<contractName>` subcell of `<account>`
```
 ...
 <account>
    <acctID> ID </acctID>
    <contractName> C </contractName>
    ...
 </account>
 ...
```

shangweilin replied:
Thanks for your information, @tomtomjhj. It really helps to solve the problem. I have revised the rules so that we match the exact/correct key. Now, the problem is solved.

dfilaretti replied:
Thanks @tomtomjhj!!

dwightguth commented on configuration.k:
```diff
@@ -0,0 +1,220 @@
+            // ---------------
+
+            <activeAccounts> .Set </activeAccounts>
+	        <cntAccount> 0:Int </cntAccount>
```
this seems redundant with the `<id>` cell above.

shangweilin replied:
Actually, `<cntAccount>` refers to the number of deployed contract instances. I will change the cell name.

ehildenb replied:
@shangweilin I'm confused about why this isn't inside the `contract` cell then. If it's tracking how many times a given contract has been deployed, should it be stored with that given contract? Or is it tracking how many contracts have been deployed total?

ehildenb replied:
But it's incremented every time that a new `<account>` is made in the rule, even if it has the same contract as another it seems. So I think it will actually just be the same as `size` of `activeAccounts`, not the number of unique contracts deployed.

dwightguth commented on configuration.k:
```diff
@@ -0,0 +1,220 @@
+                    
+                    <acctID>  .K                      </acctID>
+                    <balance> 0                      </balance>
+		            <context> .Map		     </context>
```
what is this?

shangweilin replied:
The `context` cell refers to the environment, mapping variable names to logical storage references. Actually, `context` is equivalent to the `env` cell in the `execEngine`. I will change `context` to `env` to be consistent.

shangweilin replied:
I would like to change the name to `env` to be consistent with `env` in `execEngine`, but failed. It seems that K-framework complains about having the same name. Could someone help to check? Thanks.

ehildenb replied:
@shangweilin you can't have two cells named the same thing, though it would be nice.

You could go for `acctEnv` for the ones stored in the `<account>` cell?

shangweilin replied:
I have changed the cell `<context>` into `<acctEnv>`.

dwightguth commented on configuration.k:
```diff
@@ -0,0 +1,220 @@
+                    // so that the storage usage of solidity program and evm bytecode have
+                    // the same effect on the storage.
+                    
+                    <Storage> .Map  </Storage>
```
why is this capitalized?

shangweilin replied:
Initially, we use `<storage>`, but end up with a compilation error (kompile) saying that `<storage>` conflicts with the Solidity keyword `storage`. Could you advise how to solve this issue?

ehildenb replied:
@shangweilin you won't be able to use `storage` then, I would follow the same convention and do `acctStorage`.

shangweilin replied:
Fixed.

dwightguth commented on configuration.k:
```diff
@@ -0,0 +1,220 @@
+                    //
+
+                    // not needed, maybe?
+                    <nonce>   .K             </nonce>
```
this is definitely necessary.

ehildenb replied:
Why is the nonce sort `.K`? Should it just be `0` so that it's sort `Int`?

shangweilin replied:
We are not sure if the `Int` sort in K framework is sufficient for `nonce` because the length of a nonce is 64-bit. Temporarily, we make it of the sort `.k`, just like the `acctID` cell for account ID.

tomtomjhj replied:
K's Int is unbounded so it should be ok to use them if we add some side conditions for the range of the value in the rules using nonce.

dwightguth commented on configuration.k:
```diff
@@ -0,0 +1,220 @@
+            // so that the deployed contract have access to them and can create and deploy them if needed. 
+            // (see the <account> cell below for more notes about this)
+
+            <contracts>
```
it seems strange to me that this isn't part of the network cell. shouldn't all this just live inside the `<code>` part of an account?

ehildenb replied:
I agree with @dwightguth here, the `<Contract>` subcell can be directly inside the `<code>` cell, then we'll match the other blockchain definitions better but still get the benefits of having a more structured `<Contract>` representation of the code.

tomtomjhj replied:
I think it makes sense to put `<contracts>` outside `<network>` because not all contracts are deployed.
e.g.
https://github.com/gnosis/safe-contracts/blob/6db7ce0f600d391bc6f34b144d7845bf5f227d81/contracts/common/Enum.sol
https://github.com/gnosis/safe-contracts/blob/c942db2895fa0b1003531cabb26f8dc37c999ec1/contracts/base/Executor.sol#L12
The contract `Enum` is not deployed but its code is used in another contract `Executor`.

dfilaretti replied:
The reason I suggested this approach is that I thought we'd need some temporary place to store those (processed) contract definitions before even deploying them (or using them for whatever other purpose). Let's say, for example, we load a source file; before deploying (or performing any other operation), the code has to be parsed, pre-processed (e.g. handle modifiers) and "packaged" into a data structure. Only after that, there will be a command saying "OK now deploy contract X (not the actual source, but the data structure) in address Y". So I think we need some "buffer" like that to store contracts before using them, so the idea is to only copy a contract in `<network>` if and when needed.

shangweilin replied:
Yes, exactly as @tomtomjhj  said. Not all contracts will be deployed. The network only includes all the contract instances that are deployed. So, I guess we keep `contracts` as where it is?

ehildenb replied:
Yes, leave as is for now.

ehildenb commented:
I agree with @dwightguth on pretty much all his points about the configuration. I think if you add some commits which make those changes and fix any resulting breakages in the rest of the semantics that would be good. I can take a closer look at the rules then.

ehildenb commented:
By the way, we put `type="Map"` on some cells which are `multiplicity="*"` because it make K internally treat that cell as a `Map` sort instead of a `List` sort. This means we use *the first cell* (important) as the *key* to the `Map`, so that cell *must* occur in any rule which mentions any of the cells inside `Contract`.

shangweilin commented:
I have revised the configuration accordingly as much as I can. Please proceed to review the semantics rules. Thanks.

JIAOJIAO1991 commented:
The error comes from #createStateVars(N,SV) in New-Contract-Instance. To be specific, the error comes from Allocate-State-Variables. Since a bag of the cell <account> is created in New-Contract-Instance, the subcells <context> and <Storage> need to be rewritten in #createStateVars to store state variables. new A() and new B() seem to work because they do not have any state variables to store. In other words, whenever we need to rewrite the subcells in <account> which is created in New-Contract-Instance, the error appears. It seems that rewriting of cell contents is not allowed when Map is used instead of Bag. Is it appropriate to use Map instead of Bag when rewiring of cell contents is necessary? Or how to avoid this problem when Map is used?

shangweilin commented:
I have fixed the bug of mapping incomplete keys of `Map`. Please continue reviewing the other rules.

tomtomjhj commented:
It would be nice to have a style guideline e.g. not using tabs.

tomtomjhj commented on expression.k:
```diff
@@ -0,0 +1,75 @@
+rule [Write-Local-Variables]:
+    <k> #write(X:Id,I:Value) => #end_Exp ...</k>
+    <env> ... X |-> #storedVar(N:Int, _:TypeName, #mem) ...</env>
+    <localMem> ... ((N |-> _:Value) => (N |-> I)) ...</localMem>
```
```suggestion
    <localMem> ... N |-> (_ => I) ...</localMem>
```

tomtomjhj commented on expression.k:
```diff
@@ -0,0 +1,75 @@
+    <currentAccount> #account(C, CN) </currentAccount>
+    <account>
+        <acctID> #account(C, CN) </acctID>
+        <Storage> ... ((N |-> _:Value) => (N |-> I)) ...</Storage>
```
```suggestion
    <localMem> ... N |-> (_ => I) ...</localMem>
```

JIAOJIAO1991 commented on expression.k:
```diff
@@ -0,0 +1,75 @@
+rule [Write-Local-Variables]:
+    <k> #write(X:Id,I:Value) => #end_Exp ...</k>
+    <env> ... X |-> #storedVar(N:Int, _:TypeName, #mem) ...</env>
+    <localMem> ... ((N |-> _:Value) => (N |-> I)) ...</localMem>
```
Based on my experience, sometimes the suggested one will fail. If we have something that works. Why do we have to switch to another? This is not something wrong or not standard.

tomtomjhj replied:
Do you have the error messages/minimal examples for such cases? Let's keep the current version if the suggested one is buggy.

JIAOJIAO1991 replied:
In that case, the rule cannot be applied and the program will got stuck there. But by rewriting it to the current version, it works. Sometime it happens, but sometimes not.

dfilaretti replied:
I recall we discussed this problem in the past. When you say that sometimes the problem happens but sometimes not, do you mean that it happens completely at random (e.g. running the same program N times in succession, without any other change) or that it happens after changing something apparently unrelated in the syntax/rules?

Actually it would be great if you could provide a minimal example that triggers this problem. That usually makes it easier to investigate and, if it turns out to be a K bug, to report it to the developers. 

Regarding the rule itself, I agree it's not technically wrong to use the more verbose version `((N |-> _:Value) => (N |-> I))`, but I also think the suggested version is better `N |-> (_ => I)`. 

My gut feeling is that, similarly to the previous issue with `type = Map` in `<account>`, this problem is indicating there's something incorrect somewhere, so while it may not be a priority right now, it's still worth investigating at some point.

JIAOJIAO1991 replied:
I agree that the writing is not wrong and I think it is K tool problem. You can take a look at the version in master. If you replace it with the suggested one, it cannot work. But in IMP example, it seems to work.

JIAOJIAO1991 replied:
These two expressions should be equivalent without changing other parts of the rule.

dfilaretti replied:
See https://github.com/kframework/solidity-semantics/issues/30

JIAOJIAO1991 commented on function.k:
```diff
@@ -0,0 +1,249 @@
+rule [External-Function-Call]:
+     <k> functionCall(C:Int; F:Id; Vs:Values) => #Call(F,Vs)  ~> #returnContract ...</k>
+     <currentAccount> #account(C1,N1) => #account(C,N) </currentAccount>
+     <env> Rho:Map => CRho </env>
```
Why? Do you mean deleting these two cells in the configuration which has been confirmed?

JIAOJIAO1991 commented on function.k:
```diff
@@ -0,0 +1,249 @@
+         ...
+     </Contract>
+
+rule
```
This is used to make logical relations clear.

tomtomjhj commented on function.k:
```diff
@@ -0,0 +1,249 @@
+     <currentAccount> _ => C  </currentAccount>
+     <callDepth> Depth => Depth -Int 1 </callDepth>
+
+rule [ReturnValue-In-ExternalCall]:
```
https://github.com/kframework/solidity-semantics/blob/8b967ad6b17ae8413fe204a69f0e7cb9bfa61aa4/function.k#L231-L233
https://github.com/kframework/solidity-semantics/blob/8b967ad6b17ae8413fe204a69f0e7cb9bfa61aa4/function.k#L240-L242
I think these rules are wrong because they can be applied even when the `return` is for a external call.
Try this
```solidity
contract A {
        int a;
	function g(int aa) public returns (int) {
          a = aa + 1;
          return a;
	}
}

contract B {
        int b;
	function g1() public returns (int) {
          functionCall(g2;);
          return b;
	}
        function g2() internal {
          b = functionCall(new A(); g; 123) ; // this function returns nothing
        }
}

{
"Test1" : {
	"exec" : {
	    "code" : #solidity( int ans; ans = functionCall(new B() ; g1 ; ); )
	},
        "post" : {
            "mem" : #exists("ans", 124)

        }
     }
}
```
If you krun this with `--search-all` flag, you can find
```
<localCalls>
  ListItem ( #state ( b |-> #storedVar ( 2 , int , #storage ) , #return ( true , 124 ) ) )
</localCalls>
```
and this doesn't make sense.

JIAOJIAO1991 replied:
Actually, this is not an intended behavior of external function calls.  Is it possible that external calls can be nested in local calls? Instead of calling g1, you can directly call g2. These rules are based on the assumptions that external calls cannot be nested in local calls.

JIAOJIAO1991 replied:
Btw,  functionCall(new A(); g; 123) ; and functionCall(g2;); are our internal interpretations and not valid syntax in Solidity.

tomtomjhj replied:
```solidity
contract A {
    int a;
    function g(int aa) public returns (int) {
    	a = aa + 1;
      	return a;
    }
}

contract B {
    int b;
    function g1() public returns (int) {
   	g2(); 
      	return b;
    }
    function g2() internal {
      	b = (new A()).g(123);
    }
}


```
this is a valid solidity program.

JIAOJIAO1991 replied:
If this is possible, then how can we know whether the most inside call is a local call or external call?

dfilaretti replied:
Hmm. Looks like the idea behind those rules is "if `<localCalls>` is empty then we are in an external call; otherwise we are in an internal call". But unfortunately it seems not to be that way, i.e. we can be in an external call _and_ have `<localCall>` not empty. In that case, rule `[ReturnValue-In-ExternalCall]` can't apply (because it requires an empty `<localCall>`) and rule `[ReturnValue-In-LocalCall]` applies instead, changing the content of `<localCalls>` in an unexpected way. 

I guess the question then is: how to properly distinguish between the external vs internal call cases?

dfilaretti replied:
Btw: @tomtomjhj I tried running the example with `--search-all` as you suggested, but I noticed the output is ~36K lines long. I wonder, how did you find out about that? Did you manually inspect all the output or did you search for something?

JIAOJIAO1991 replied:
The previous assumption is that locals can be nested in external calls, but external calls cannot be nested in local calls. However, they can be present at the same time. If we use one stack for all function calls, the problem will be solved.

tomtomjhj replied:
possible solutions
* just don't use `<localCalls>`. Put everything in `<callStack>` and put some makers on the elements
* 
```
rule [External-Function-Call]:
      ......
     <localCalls> LocalCallStack => .List </localCalls>
     <callStack> .List => ListItem(#state(#account(C1,N1), Rho, LocalCallStack, #return(false,#undef_Value))) ... </callStack>
     ......

rule [Return-In-ExternalCall]:
     <k> return  => #end_Exp ... </k>
     <localCalls> .List => LocalCallStack </localCalls>
     <callStack> ListItem(#state(C, Rho, LocalCallStack #return(_:Bool => true, _:Value))) ... </callStack>
......
```
with these modifications, a `return` is for internal func iff `<localCalls>` is not empty.

tomtomjhj replied:
@dfilaretti I first suspected that there might be some issues with this, wrote/ran the example code, and just used vim to find that line.

tomtomjhj replied:
@JIAOJIAO1991 agreed.

dfilaretti replied:
So which approach have you decided to take? 

At a first look, both possible solutions proposed by by @tomtomjhj look reasonable to me. 
Maybe solution (2) will require less changes to the current code (since you get to keep both stacks)? But at the same time solution (1) requires 1 less cell and is probably simpler (you just add some tag to the stack frame and look at that to decide whether it's internal or external). 

I'm not sure what's better though. 
It would be great to hear someone else's thoughts on this. @dwightguth @ehildenb do you have any advice to share? Thank you!

shangweilin replied:
I suggest option one, i.e., using only one stack `callStack` with addition informative markers to distinguish between external or internal function calls.

The reason is that a stack is supposed to help us to keep track of the thing happens most recently (last in first out). If we use two stacks, then we lose the order relation of elements between two stacks, that is, which one happens the most recently, the top of stack 1 or the top of stack 2? Of course we can put some addition effort to maintain the order/sequence, but it somehow complicates the flow.

What do you think? @dwightguth, @ehildenb, and @tomtomjhj.

ehildenb replied:
@shangweilin what about an extra cell inside `callState` marking if it's a local or external call? Similar to how the `static` cell works, it will just be a single `Bool` that you need to check.

Then, when you do `#pushCallStack`, because the entire contents of `callState` will be saved, it will also save off that particular value.

shangweilin replied:
We have fixed this issue by removing the cell `localCalls`. That is, we only have one stack now, which is `<callStack>`. In addition, we add `FunctionFlag`, which could be `#internal` or `#external`, into each stack frame, so that we can distinguish between external/local function calls.

ehildenb replied:
Marking as resolved to keep conversations concise. I have additional comments below about this subject that I think should be resolved.

dfilaretti commented:
I was looking once again at the compiler tests (https://github.com/ethereum/solidity/blob/develop/test/libsolidity/SolidityEndToEndTest.cpp) and noticed how a simple search for `function_` shows many possible test cases that you could actually start using (rather than having to write your own). 

Like this for example: 

```
BOOST_AUTO_TEST_CASE(same_function_in_construction_and_runtime)
{
	char const* sourceCode = R"(
		contract C {
			uint public initial;
			constructor() public {
				initial = double(2);
			}
			function double(uint _arg) public returns (uint _ret) {
				_ret = _arg * 2;
			}
			function runtime(uint _arg) public returns (uint) {
				return double(_arg);
			}
		}
	)";

	compileAndRun(sourceCode, 0, "C");
	ABI_CHECK(callContractFunction("runtime(uint256)", encodeArgs(u256(3))), encodeArgs(u256(6)));
	ABI_CHECK(callContractFunction("initial()"), encodeArgs(u256(4)));
}
```

or this: 

```
BOOST_AUTO_TEST_CASE(pass_function_types_internally)
{
	char const* sourceCode = R"(
		contract C {
			function f(uint x) public returns (uint) {
				return eval(g, x);
			}
			function eval(function(uint) internal returns (uint) x, uint a) internal returns (uint) {
				return x(a);
			}
			function g(uint x) public returns (uint) { return x + 1; }
		}
	)";

	compileAndRun(sourceCode, 0, "C");
	ABI_CHECK(callContractFunction("f(uint256)", 7), encodeArgs(u256(8)));
}
```

I know many of those are not supposed to pass (either because they make use of features you do not support yet, or because there are issues with the current definition), but that's exactly the point of those tests: they give us a goal and a way to figure out when we are "done".

tomtomjhj commented:
@dfilaretti tests that only use `ABI_CHECK` could be automated with K by defining those (simple) CPP functions in K. But definitely `testContractAgainstCppOnRange` stuff will be nontrivial.

dfilaretti commented:
well my comment was only meant to suggest some possible tests to look at, even if we don't have a way to automate it yet. Even if done manually, it can still give us something to play with for the time being. But I agree, in the long term we need to automate, as it's impractical to rewrite them all manually (and keep them up to date as the language evolves).

shangweilin commented:
I have added the desugaring rules for internal function calls. That is, for internal function calls, we can use exact Solidity syntax now, instead of `functionCall`.

ehildenb commented on contract.k:
```diff
@@ -0,0 +1,112 @@
+         <stateVar> SV:List </stateVar>
+      ...
+     </Contract>
+     <numOfAccounts> N:Int => N +Int 1 </numOfAccounts>
```
It looks like we're tracking the total number of accounts deployed here, but *not* the set of `<activeAccount>`. Perhaps instead of adding cell `<numOfAccounts>`, just update the cell `<activeAccount>`. Something like:

```
<activeAccounts> ... (.Set => SetItem(#account(N, C))) </activeAccounts>
```

Then if we ever need the number of accounts, we can always call `size` on the contents of `activeAccounts`.
```

shangweilin replied:
I have managed to figure out the syntax of call `size` on `activeAccounts`. This issue is fixed accordingly.

ehildenb replied:
@shangweilin you can also look at file `k-distribution/include/builtin/domains.k` in the `kframework/k` distribution to see how these builltins are defined.

ehildenb commented on configuration.k:
```diff
@@ -0,0 +1,228 @@
+                <callStack>     .List  </callStack>
+                <interimStates>   .K  </interimStates>
+                <touchedAccounts> .K  </touchedAccounts>
+                <currentAccount>  #account(#undef_Int,#undef_Id)  </currentAccount>
```
This is the only place where `#undef_Int` and `#undef_Id` are used. I would recommend instead creating a value `.Account` or `#undef_Account` like so:

```
syntax AccountId ::= Int | "#undef_Account"
```

You declare `#undef_Int` for example here:

```
syntax Int ::= "#undef_Int"
```

which means you're adding junk to the `Int` sort, which will make reasoning using unification and Z3 very difficult when it comes time to do the proofs. Almost always better to use a super-sort (like the `AccountId` super-sort above) and add the extra constants there.

ehildenb replied:
@shangweilin is it ever the case that it's useful to have an `#account` with only one of the entries undefined? Eg: `#account(#undef_Int, x)` or `#account(3, #undef_Id)`? If not (and I can't see any uses in the semantics of these cases), then instead of adding sorts `AccountId` and `AccountInt` as you have, I would just add an extra constant `#undef_Account` to sort `Account`.

```
syntax Account ::= #account(Int, Id) | "#undef_Account"
```

If the semantics ever could produce examples like above, then it would make sense, but otherwise would be better to avoid adding super-sorts of builtin sorts (like `Int` and `Id`) if possible.

Note that this pattern of adding an extra constant to a sort (like `#undef_Account`) is like the `Maybe` type in Haskell, or the `Optional` in Java/Scala.

ehildenb commented on configuration.k:
```diff
@@ -0,0 +1,228 @@
+                <statusCode>      .K  </statusCode>
+                <callStack>     .List  </callStack>
+                <interimStates>   .K  </interimStates>
+                <touchedAccounts> .K  </touchedAccounts>
```
Why are `<callStack>` and `<interimStates>` starting with `.K` instead of `.List`? In the semantics you use them like they are `List`.

ehildenb commented on configuration.k:
```diff
@@ -0,0 +1,228 @@
+                    // Gas: not sure how, but I think we need it. 
+                    <gas>         .K          </gas>
+                    <memoryUsed>  .K          </memoryUsed>
+                    <previousGas> .K          </previousGas>
```
These should all start with `0`, not with `.K`.

ehildenb commented on configuration.k:
```diff
@@ -0,0 +1,228 @@
+
+                    // Those are present in both KEVM and IELE. 
+                    // Not sure we need them, let's keep them around until figure it out. 
+                    <static>    .K     </static>
```
This should start as `false`, not as `.K`. You will need this eventually, it marks whether we're in a call which is not allowed to change the storage state of any accounts.

shangweilin commented on README.md:
```diff
@@ -0,0 +1,20 @@
+
+3. If the test case is passed then the cell <mode/> in the configuation is success else
+   the cell <k/> is stuck at some unfinished term.
+
```
@dfilaretti, I still have the following warning message
```
[Warning] Critical: Found 0 integer variables in exit code pattern. Returning
111
```

Do you know how to get rid of it?

ehildenb replied:
Start the `<exit-code>` cell with a `0` instead of a `1` in it if you want to disable it completely.

But really, what should happen is when a test passes it should set `<exit-code>` to zero, so that you know it passed. This is how KEVM does it. So basically, there is a rule which goes something like:

```
<k> success => . ... </k>
<exit-code> _ => 0 </exit-code>
```

ehildenb replied:
@shangweilin it looks like the exit-code cell hasn't been started with `0`, and is not being used yet.

Eventually, it's much quicker to run tests using the exit-code as a success metric than by comparing output (because then you need to open a file on disk and write to it), which will be needed once we are executing lots of tests. For now, I think just set it to `0` in the configuration so that it's not printing the annoying error message.

shangweilin commented:
Hi, @dfilaretti and @ehildenb. We have addressed all the comments except for the `numOfAccounts` because we don't know the syntax of apply `size` (is it a function or an attribute?) on `<activeAccounts>`, which is a set. Please advise, thanks.

shangweilin commented:
Hi, @dfilaretti, @ehildenb, and @dwightguth, I managed to figure out the usage of `size`. I think we have fixed all the issues that you raised. Please hep to review the code again, thanks.

ehildenb commented on function.k:
```diff
@@ -0,0 +1,231 @@
+
+rule [Return]:
+     <k> return  => #end_Exp ... </k>
+     <callStack> ListItem(#state(C, Rho, #return(_:Bool => true, _:Value), _:FunctionFlag)) ... </callStack>
```
I think it would make more sense to refer to the values in the *current* `<callState>` rather than looking at the top of the `<callStack>` for making these decisions.

ehildenb commented on function.k:
```diff
@@ -0,0 +1,231 @@
+     <k> functionCall(F:Id; Vs:Values) => #Call(F,Vs) ~> #return ... </k>
+     <currentAccount> #account(C,N) </currentAccount>
+     <env> ENV:Map  </env>
+     <callStack> .List => ListItem(#state(#account(C,N),ENV,#return(false,#undef_Value), #internal)) ... </callStack> 
```
I think the `<callStack>` should be modeled more directly off of the KEVM one, so that when you push to the `<callStack>` all you have to do is put a copy of the entire `<callState>` sub-configuration onto the front of the stack. It makes these rules very simple, and makes it so that it's easy to decide whether a part of the configuration belongs in the `<callState>` sub-configuration: If each call-frame (ie. elements of the `<callStack>`) needs its own copy of a piece of state then add that piece of state inside the `<callState>` sub-configuration, if not then put it outside that sub-configuration.

dfilaretti commented on expression.k:
```diff
@@ -0,0 +1,75 @@
+    <currentAccount> #account(C, CN) </currentAccount>
+    <account>
+        <acctID> #account(C, CN) </acctID>
+        <Storage> ... ((N |-> _:Value) => (N |-> I)) ...</Storage>
```
See #30

dfilaretti commented on tests/functionCall.sol:
```diff
@@ -0,0 +1,55 @@
+			sum = 0;
+		}
+		else {
+			sum = x + functionCall(f ; x - 1);
```
This is done for internal calls. 
Later on, it should also be done for external calls. 
To keep track of it, I created an issue: https://github.com/kframework/solidity-semantics/issues/35.

shangweilin commented:
Hi, @ehildenb, @dfilaretti, @tomtomjhj, and @dwightguth,  I have fixed the final three issues. Please help to review again, thanks.

ehildenb commented:
@shangweilin there is still a lot to be improved, but I'll submit a follow-up PR which fixes the formatting issues and some other things. Feel free to merge ASAP.
